jctools-core/src/main/java/org/jctools/queues/MpscArrayQueue.java

```diff
@@ -331,32 +338,78 @@ public E poll()
         // Copy field to avoid re-reading after volatile load
         final E[] buffer = this.buffer;
 
-        // If we can't see the next available element we can't poll
+        long seenPIndex = seenProducerIndex;
         E e = lvElement(buffer, offset); // LoadLoad
-        if (null == e)
+        if (e != null)
         {
-            /*
-             * NOTE: Queue may not actually be empty in the case of a producer (P1) being interrupted after
-             * winning the CAS on offer but before storing the element in the queue. Other producers may go on
-             * to fill up the queue after this element.
-             */
-            if (cIndex != lvProducerIndex())
+            incrementConsumerIndex(buffer, offset, cIndex, seenPIndex);
+            return e;
+        }
+        if (seenPIndex <= cIndex && (seenPIndex = lvProducerIndex()) == cIndex)
+        {
+            return null;
+        }
+        return pollMissPath(buffer, offset, cIndex, seenPIndex);
+    }
+
+    private void incrementConsumerIndex(final E[] buffer, long offset, long cIndex, final long pIndex) {
+        // This is called after the current cIndex slot has been read, and moves the consumer index
+        // to the next non-CONSUMED slot, nulling all preceding slots
+        spElement(buffer, offset, null);
+        // If no lookahead has happened then pIndex <= cIndex + 1 and we won't enter the loop
+        while (++cIndex < pIndex && lpElement(buffer, offset = calcElementOffset(cIndex)) == CONSUMED)
+        {
+            spElement(buffer, offset, null);
+        }
+        soConsumerIndex(cIndex);
+    }
+
+    private E pollMissPath(final E[] buffer, final long cOffset, final long cIndex, long pIndex)
+    {
+        while (true)
+        {
+            for (int i = 0; i < SPIN_COUNT; i++)
             {
-                do
+                E e = tryRange(buffer, cOffset, cIndex, pIndex);
+                if (e != null)
                 {
-                    e = lvElement(buffer, offset);
+                    seenProducerIndex = pIndex;
+                    return e;
                 }
-                while (e == null);
             }
-            else
+            pIndex = lvProducerIndex();
```

Comment:
I'm not sure I would query the producer index SPIN_COUNT times, but just once.
It would harm the AtomicXXX/future VarHandle version because bounds checking will be performed each time + I would try to maintain control over the maximum stride/distance of checked elements to save the cache miss while checking back the first element (so will try to be 2 cache lines/oops size elements).

Reply:
`SPIN_COUNT` is the number of iterations _between_ producer index reads, this is to attenuate producer contention without preventing us from discovering more distant produced elements that might become visible sooner.
> It would harm the AtomicXXX/future VarHandle version because bounds checking will be performed each time

Please elaborate... which bounds checking? :)

> I would try to maintain control over the maximum stride/distance of checked elements

It seems like this could be sensible, but feel like it may be best to determine that empirically.

Reply:
Uuups I have misread parentheses :P
Each time the producer index change, the bound checks injected by the JVM on the tryRange loop cannot be hoisted out until the outer loop (on pollMissPath) if tryRange will be inlined, but will be performed each time.
Although it should be the "slow path", we don't know when any offered element will appear, so making such loop as fast as possible help to guarantee good latencies. 
If tryRange won't be inlined due to its size I suggest to inline it manually too.
These are just some "premature" optimizations, but nonetheless seems to me plausible and easy to be achieved I'd won't break the algorithm